src/abc_atlas_access/abc_atlas_cache/cloud_cache.py

```diff
@@ -0,0 +1,1137 @@
+        ----------
+        manifest_name: str
+            The name of the manifest to load. Must be an element in
+            self.manifest_file_names
+        """
+        response = self.s3_client.get_object(Bucket=self._bucket_name,
+                                             Key=manifest_name)
+
+        filepath = self._cache_dir / manifest_name
+        filepath.parents[0].mkdir(parents=True, exist_ok=True)
+
+        with open(filepath, 'wb') as f:
+            for chunk in response['Body'].iter_chunks():
+                f.write(chunk)
+
+    def _download_file(self, file_attributes: CacheFileAttributes) -> bool:
+        """
+        Check if a file exists locally. If it does not, download it
+        and return True. Return False otherwise.
+
+        Parameters
+        ----------
+        file_attributes: CacheFileAttributes
+            Describes the file to download
+
+        Returns
+        -------
+        bool
+            True if the file was downloaded; False otherwise
+
+        Raises
+        ------
+        RuntimeError
+            If the path to the directory where the file is to be saved
+            points to something that is not a directory.
+
+        RuntimeError
+            If it is not able to successfully download the file after
+            10 iterations
+        """
+        was_downloaded = False
+
+        local_path = file_attributes.local_path
+
+        local_dir = local_path.parents[0]
+        local_dir.mkdir(parents=True, exist_ok=True)
+        if not os.path.isdir(local_dir):
+            raise RuntimeError(f"{local_dir}\n"
+                               "is not a directory")
+
+        obj_key = file_attributes.relative_path
+
+        n_iter = 0
+        max_iter = 10  # maximum number of times to try download
+
+        pbar = None
+        if not self._file_exists(file_attributes):
+            response = self.s3_client.list_objects_v2(Bucket=self._bucket_name,
+                                                      Prefix=str(obj_key))
+            object_info = response['Contents'][0]
+            pbar = tqdm.tqdm(desc=object_info["Key"].split("/")[-1],
+                             total=object_info["Size"],
+                             unit_scale=True,
+                             unit_divisor=1000.,
+                             unit="MB")
+
+        while not self._file_exists(file_attributes):
+            was_downloaded = True
+            response = self.s3_client.get_object(Bucket=self._bucket_name,
+                                                 Key=str(obj_key))
+
+            if 'Body' in response:
+                with open(local_path, 'wb') as out_file:
+                    for chunk in response['Body'].iter_chunks():
+                        out_file.write(chunk)
+                        pbar.update(len(chunk))
+
+            # Verify the size of the downloaded file
+            if file_attributes.file_size != file_attributes.local_path.stat().st_size:
```

Comment:
Yeah, forgot that I wanted to change this (was testing withe latest manifest). The size does work for that one but, as I discovered due to this check, the oldest manifest has an incorrect size.

As you mentioned previously, that is an option to add hashes to the manifests so that we can confirm the file downloaded properly and has not been changed on disk. Another option is to just use the size in the manifest and on disk to check and return a warning that the file may have downloaded imporperly and to check the results, having a "force_download" option as well.

The safest is to add a hash to the manifest. That way we would be able to confirm the downloaded file and reuse the internal cache manifest from the previous version. I don't think we really need the latter due to how there's no versioning in this repo. The main thing to keep the hash around is verifying the file integrity.

Reply:
I suspect people will be uneasy about updating the manifest files that have already been released (though I agree that it is the safest option).

In the SDK, we tried to make the API very smart. Users instantiated a VisualBehaviorOphysCache and called get_session(session_id=....) and the data was automatically downloaded if it wasn't there.

What if, in this case, we don't make everything so automatic. When users ask to load a data file, the cache object (or whatever they are looking for) can just check if the file exists and, if it doesn't, say "You don't have this file. Download it with...." If the downloads are not triggered automatically, we can change the download_file() method to download the file even if it already exists. It is then on the user not to ask to download the same file more than once.

File integrity is also, in this case, the user's problem. They have to trigger a download if the file ends up being corrupted (I strongly expect that if the data is improperly downloaded, they will not be able to manipulate the data with anndata).

Reply:
Put the hashing stuff in the code (and updated some manifests locally to test). Added a force_download and skip hashing as we discussed with Lydia.